diesel_cli/src/infer_schema_internals/inference.rs

```diff
@@ -23,6 +23,38 @@ fn is_reserved(name: &str) -> bool {
         )
 }
 
+fn contains_unmappable_chars(name: &str) -> bool {
+    // Rust identifier names are restricted to [a-zA-Z0-9_].
+    !name.chars().all(|c| c.is_ascii_alphanumeric() || c == '_')
+}
+
+pub fn rust_name_for_sql_name(sql_name: &str) -> Option<String> {
```

Comment:
Can we return a `Cow<str>` here instead of the option? That would remove all that `.unwrap_or(…)` below.

Reply:
Unfortunately, a None variant returned here will be checked in `print_schema.rs` (L206) to stop diesel_cli from adding the `#[sql_name = "..."]` attribute macro header in case the SQL name is a valid Rust identifier. I'm not sure it'd be straightforward to swap out the Option here.

Reply:
(I suppose we could use `Cow::is_borrowed()` for that check, but unfortunately it's nightly-only for now)

Reply:
We should be able to detect such cases by just comparing `sql_name` with `rust_name`. If they are not equal we need the attribute.

Reply:
Of course, you're right. I'll make the changes!

Reply:
So, my thinking was that `rust_name: Cow<str>` would either point to an owned string, meaning `sql_name` had been modified, or to `&sql_name`, meaning it was unchanged. Then it'd be a simple ` if let Cow::Owned(_) = rust_name {}` to know whether we needed to add the attribute header to the schema or not.

However, storing `rust_name` as Cow in `data_structures::ColumnDefinition` forces me to add a lifetime to it, which ends up polluting _a lot_ of the code in diesel_cli with lifetime annotations. It's not very elegant.

I then tried to store `rust_name` as a String and do the comparison as you suggested ([see this commit](https://github.com/abreis/diesel/commit/964d8e62aa86963fffe18759ee1c9944d1ed5bf0)), converting from Cow\<str\> where needed. But now I'm doing string comparisons instead of checking for an Option variant, so I'm not sure this is better.

Thoughts?

Reply:
I think that variant look better than using `Option<String>`. As `Cow<str>` would introduce lifetimes I would say go for `String` instead in all places. In the end this is no performance critical code, so just allocating the string there for having a cleaner code is fine.